src/connection.c

```diff
@@ -685,18 +688,30 @@ int createNopollConnection(noPollCtx *ctx)
 	conn_ctx.nopoll_ctx = ctx;
 	init_expire_timer (&conn_ctx.connect_timer);
 	init_header_info (&conn_ctx.header_info);
-        set_server_list_null (&conn_ctx.server_list);
-        init_backoff_timer (&backoff_timer, max_retry_count);
+	/* look up server information if we don't already have it */
+	if (server_is_null (&server_list->defaults))
+	  if (find_servers (server_list) == FIND_INVALID_DEFAULT) {
+        return nopoll_false;		  
+      }
+	conn_ctx.server_list = server_list;
+    init_backoff_timer (&backoff_timer, max_retry_count);
+    start_conn_in_progress (backoff_timer.start_time); 
   
 	while (!g_shutdown)
 	{
-	  query_dns_status = find_servers (&conn_ctx.server_list);
-	  if (query_dns_status == FIND_INVALID_DEFAULT)
-		return nopoll_false;
 	  set_current_server (&conn_ctx);
 	  if (keep_trying_to_connect (&conn_ctx, &backoff_timer))
 		break;
-	  // retry dns query
+	  /* if we failed to connect, don't reuse the redirect server */	
+      free_server (&conn_ctx.server_list->redirect);
+#ifdef FEATURE_DNS_QUERY
+      /* if we don't already have a valid jwt, look up server information */
+      if (server_is_null (&conn_ctx.server_list->jwt))
+        if (find_servers (conn_ctx.server_list) == FIND_INVALID_DEFAULT) {
+		  /* since we already found a default server, we don't expect FIND_INVALID_DEFAULT */
+		  g_shutdown = true;
```

Comment:
Why is shutdown called here?

Reply:
find_servers should not get an INVALID_DEFAULT return.
We already succeeded once.
If we do get it, we need to shut down in any case.